### PR TITLE
Allow to apply network disruptions on init containers

### PR DESCRIPTION
**Pull request by Devatoria**
## What does this PR do?

- [x] Adds new functionality
- [x] Alters existing functionality
- [ ] Fixes a bug
- [ ] Improves Documentation

A brief description of changes to implementation or controller behavior:

This PR adds a mutating webhook in the admission webhooks which inserts an init container in pods having a specific label. This containers is inserted as the very first init container and prevents other containers from being created, waiting for a `SIGUSR1` signal to exit properly. Because the network namespace is shared across all containers within a pod, this container allows a disruption to be injected before any other containers are created in the targeted pod.

Motivation for change:

It allows us to apply a (network) disruption on containers before they are created and started so we can disrupt startup dependencies easily.

### Testing

- [x] I used existing unit tests
- [x] I manually tested locally
- [x] I will manually test in a canary deployment

Please list your manual testing steps (sample `yaml` file, commands, important checks):

### Checklist

- [x] The documentation is up to date
- [x] My code is sufficiently commented
- [x] I have tested to the best of my ability
- [x] I have signed my commit (see [Contributing Docs](../CONTRIBUTING.md))

### CODE REVIEW

File: webhook/chaos_handler.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+)
+
+type ChaosHandlerMutator struct {
+	Client  client.Client
+	Log     *zap.SugaredLogger
+	Image   string
+	decoder *admission.Decoder
+}
+
+func (m *ChaosHandlerMutator) InjectDecoder(d *admission.Decoder) error {
```
Comment (ptnapoleon) [inline webhook/chaos_handler.go:26]: Who calls this? I wasn't able to find any usages

File: webhook/chaos_handler.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+	m.decoder = d
+
+	return nil
+}
+
+func (m *ChaosHandlerMutator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	pod := &corev1.Pod{}
+
+	err := m.decoder.Decode(req, pod)
```
Comment (ptnapoleon) [inline webhook/chaos_handler.go:35]: `m.decoder` can be nil here, should we check that?

File: cli/handler/main.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	"github.com/spf13/cobra"
+	"go.uber.org/zap"
+)
+
+var (
+	logger *zap.SugaredLogger
+
+	rootCmd = &cobra.Command{
+		Use:   "cobra",
+		Short: "A generator for Cobra based Applications",
```
Comment (ptnapoleon) [inline cli/handler/main.go:24]: Just a placeholder so we remember to change this

File: docs/features.md [hunk 4]
```diff
@@ -32,6 +32,23 @@ How can you target a specific pod by name, if it doesn't have a unique label sel
...
+
+> :memo: This mode has some restrictions:
+>   * it only works for network related (network and dns) disruptions
+>   * it only works with the pod level
+>   * it does not support containers scoping (applying a disruption to only some containers)
+
+It can be handy to disrupt packets on pod initialization, meaning before containers are actually created and started, to test startup dependencies or init containers. You can do this in only two steps:
+
+* redeploy your pod with the specific label `chaos.datadoghq.com/disrupt-on-init` to hold it in the initialization state
+  * the chaos-controller will inject an init containers named `chaos-handler` in your pod
```
Comment (ptnapoleon) [inline docs/features.md:45]:
```suggestion
  * the chaos-controller will inject an init containers name `chaos-handler` as the first init container in your pod
```

File: docs/features.md [hunk 5]
```diff
@@ -32,6 +32,23 @@ How can you target a specific pod by name, if it doesn't have a unique label sel
...
+  * this init container is lightweight and does nothing but waiting for a `SIGUSR1` signal to complete successfully
+* apply your disruption [with the init mode on](../examples/on_init.yaml)
```
Comment (ptnapoleon) [inline docs/features.md:47]: It works if the disruption is applied first, yes? As the disruption wont have any targets to pick yet and will wait for the applicable pods to appear

File: webhook/chaos_handler.go [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	// define pod name for logs
+	// if the pod is created from a replicaset (deployment, statefulset)
+	// the name won't be populated yet because it relies on the name generation
+	podName := pod.Name
+	if podName == "" {
+		podName = pod.ObjectMeta.GenerateName
```
Comment (ptnapoleon) [inline webhook/chaos_handler.go:57]: This is only going to be a prefix of the full pod name. That's fine because we're only using it to log, afaict, but is that worth mentioning explicitly in the comments?

File: webhook/chaos_handler.go [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+		ImagePullPolicy: corev1.PullIfNotPresent,
+		Args: []string{
+			"--timeout",
+			m.Timeout.String(),
+		},
+	}
+
+	// prepend chaos handler init container to already existing init containers
+	inits := []corev1.Container{}
+	inits = append(inits, init)
```
Comment (ptnapoleon) [inline webhook/chaos_handler.go:75]: Why choose to do this over two lines rather than `inits = append(inits, init, pod.Spec.InitContainers...)`?

**Review by ptnapoleon — APPROVED** [review]

File: webhook/chaos_handler.go (hunk 7 above)
Comment (Devatoria) [reply webhook/chaos_handler.go:75]: `append` can't be used with this syntax in Go ([this answer](https://stackoverflow.com/questions/18946837/go-variadic-function-and-too-many-arguments) for more explanations). However I can indeed shorten this to a single line with another syntax, thanks for the heads up.
